### PR TITLE
New add custom basemap modal w/ XYZ URLs

### PR DESCRIPTION
**Pull request by viddo**
change-basemap

This implement the base for the [add-custom-basemaps + 1st tab w/ XYZ URLs use-case](3633) #3633

- A lot of UI states due to the async layer creation and URL validation, that was extracted/refactored on the old models.
- Implementing the rest of the use-cases should be easier now that the base is in place
  - basically just adding a view+model for each tab, as done with the XYZ use-case here.
- Maintainted existing logic and handling for basemaps in terms of error cases and validation

### CODE REVIEW

**Comment by viddo** [comment]
@xavijam @javierarce please code review
cc @saleiva

File: lib/assets/javascripts/cartodb/models/map.js [hunk 1]
```diff
@@ -249,8 +249,53 @@ cdb.admin.TileLayer = cdb.geo.TileLayer.extend({
...
+   * @param {String} url
+   * @return {cdb.admin.TileLayer}
+   */
+  byCustomURL: function(url) {
+    url = url.replace(/\{S\}/g, '{s}')
+      .replace(/\{X\}/g, '{x}')
+      .replace(/\{Y\}/g, '{y}')
+      .replace(/\{Z\}/g, '{z}');
+
+    if (url.indexOf('{x}') === -1) {
```
Comment (javisantana) [inline lib/assets/javascripts/cartodb/models/map.js]: a valid z/x/y url could be http://test.com/image.png, it does not need to have {z}, {x} or {y}. In this case a warning could be ok

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/3738/))

File: lib/assets/javascripts/cartodb/models/map.js (hunk 1 above)
Comment (viddo) [inline lib/assets/javascripts/cartodb/models/map.js]: Aha! This logic was extracted from [the old view](https://github.com/CartoDB/cartodb/blob/master/lib/assets/javascripts/cartodb/table/basemap/basemap_dialog/zxy_basemap_pane.js#L60-L74), I didn't want to change that before the review. I'll relax these checks then :+1:

**Comment by saleiva** [comment]
Ideally, the error should displace the button, but is a nice-to-have

File: lib/assets/javascripts/cartodb/common/dialogs/add_custom_basemap/xyz_view.js [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+            self._indicateIsValidating(false);
+            self._updateError();
+          }
+        },
+        error: function() {
+          if (thisCallSeq === self._lastCallSeq) {
+            self._disableOkBtn(false);
+            self._indicateIsValidating(false);
+            // Note that this text can not be longer, or it will exceed available space of the error label.
+            self._updateError("We couldn't validate this, if you're sure it contains data click \"set basemap\"");
```
Comment (xavijam) [inline lib/assets/javascripts/cartodb/common/dialogs/add_custom_basemap/xyz_view.js]: Could this error be in the template?

**Comment by xavijam** [comment]
Looks good to me

**Comment by javierarce** [comment]
:+1: :balloon:

File: lib/assets/javascripts/cartodb/common/dialogs/add_custom_basemap/xyz_view.js (hunk 2 above)
Comment (viddo) [inline lib/assets/javascripts/cartodb/common/dialogs/add_custom_basemap/xyz_view.js]: There may be different error messages (this is just one of them), moving this to the template would mean duplicating markup for each error that may happen which I don't think that makes a lot of sense.

Comment (xavijam) [inline lib/assets/javascripts/cartodb/common/dialogs/add_custom_basemap/xyz_view.js]: up to you :)

**Comment by viddo** [comment]
@saleiva not sure that's easily done, reusing the styles for existing errors. If you want to change that we should probably revisit the other places where it's used so it's consistent

**Comment by saleiva** [comment]
exactly. Leave it for the future :)

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/3759/))

**Comment by viddo** [comment]
Okok. Will do the TMS option separately too

**Pull request merged**